snapcraft/plugins/kernel.py

```diff
@@ -355,11 +379,78 @@ def _copy_dtbs(self):
             for f in found_dtbs:
                 os.link(f, os.path.join(dtb_dir, os.path.basename(f)))
 
+    def _do_parse_config(self, config_path):
+        # tokenize .config and store options in builtin[] or modules[]
+        with open(config_path) as f:
+            for line in f:
+                tok = line.strip().split('=')
+                items = len(tok)
+                if items == 2:
+                    opt = tok[0].upper()
+                    val = tok[1].upper()
+                    if val == 'Y':
+                        builtin.append(opt)
+                    elif val == 'M':
+                        modules.append(opt)
+
+    def _do_check_config(self):
+        # check the resulting .config has all the necessary options
+        msg = 'Your kernel config is missing some feature that ubuntu core \
+recommend / requires,\nand while we won\'t prevent you from building this\
+kernel snap,\n we suggest you take a look at these:'
+        required_opts = required_generic + required_security + \
+            required_snappy + required_systemd
+        missing = []
+
+        for code in required_opts:
+            opt = 'CONFIG_{}'.format(code)
+            if opt in builtin:
+                continue
+            elif opt in modules:
+                continue
+            else:
+                missing.append(opt)
+
+        if missing:
+            logger.warn('\n{}\n'.format(msg))
+            for opt in missing:
+                logger.warn('  - {}'.format(opt))
+            logger.warn('')
+
+    def _do_check_initrd(self):
+        # check all required_boot[] items are either builtin or part of initrd
+        msg = 'The following feature is boot essential, consider making\n\
+it static or adding the relevant module to initrd:'
+        missing = []
+
+        for code in required_boot:
+            opt = 'CONFIG_{}'.format(code.upper())
+            if opt in builtin:
+                continue
+            elif opt in modules:
+                if code in self.options.kernel_initrd_modules:
+                    continue
+                else:
+                    missing.append(opt)
+
+        if missing:
+            logger.warn('\n{}\n'.format(msg))
+            for opt in missing:
+                logger.warn('  - {}'.format(opt))
+            logger.warn('')
+
     def pull(self):
         super().pull()
         snapcraft.download(
             'ubuntu-core', 'edge', self.os_snap, self.project.deb_arch)
 
+    def build(self):
+        super().prep_build()
+        self._do_parse_config(self.get_config_path())
+        self._do_check_config()
+        self._do_check_initrd()
+        super().finish_build()
```

Comment:
Following my suggestions above, this could look like:
```

def build(self):
   self.prepare_config()
   builtin, modules = self.parse_config()
   self.check_config(builtin, modules)
   self.check_initrd(builtin, modules)
   self.do_build()
   self.do_install()
```